File: clojure/p26.clj
(ns p26
  (:use [euler :only (max-key timed-test)]))

;; Get period of the repeating decimal 1/d by repeated long division
;; Track remainders we see along the way
;; When a remainder is seen for the second time, we can compute the period
(defn period
  ([d] (period 1 d []))
  ([n d remainders]
     (let [r (rem (* n 10) d)]
       (cond
	(zero? r) 0
	(some #{r} remainders) (count (drop-while #(not= r %) remainders))
	:else (recur r d (conj remainders r))))))

(timed-test
 983
 (apply max-key period (range 1 1000)))

File: cpp/p26.cpp
#include "euler.h"

using namespace std;

int get_period( int a, int d, vector<int>& remainders ) {
  if( a == 0 ) 
    return 0;
	
  int remainder = ( a < d ) ? a : a % d;
	
  vector<int>::iterator i = find( remainders.begin(), remainders.end(), remainder );
	
  if( i != remainders.end() )
    return remainders.end() - i;
	
  remainders.push_back( remainder );

  return get_period( remainder * 10, d, remainders );
}

int main() {
  int d = -1;
  int longest_period = -1;

  for( int i = 1; i < 1000; i++ ) {
    vector<int> remainders;
    int period = get_period( 1, i, remainders );

    if( period > longest_period ) {
      d = i;
      longest_period = period;
    }
  }
	
  cout << d << endl;
}

